docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,573 @@
+Only the spec will be copied from the running pod. Copying metadata such as
+labels might result in the pod copy receiving traffic or being killed
+immediately by a replication controller. It would be trivial to provide a
+`--copy-labels` option if this is desired behavior.
+
+### Running Debug
+
+*Running Debug* requires new functionality in the kubelet and a new `kubectl
+debug` command that results in a new container being introduced in the running
+pod. This *debug container* is not part of the pod spec, which remains
+immutable, and is not restarted according to the pod's `restartPolicy`.
+
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
+field of `PodStatus`, which is a read-only list of `ContainerStatus` that
+contains an entry for every *debug container* that has ever run in this pod, and
+is reported by `kubectl describe`. This mirrors how `InitContainers` are
+implemented.
+
+Debug operations will generate an event so auditing utilities can reconstruct
+what commands are run, even though `kubectl exec` does not do this. A *Debug
+Containers* section will display all debug containers in the output of `kubectl
+describe` similar to the *Init Containers* section.
+
+The following command would attach to a newly created container in a pod:
+
+```
+kubectl debug -it -c debug-shell --image=debian target-pod -- bash
+```
+
+It would be reasonable for Kubernetes to provide a default container image
+which, combined with a default entrypoint, makes the minimal debug command:
+
+```
+kubectl debug -p target-pod -it
+```
+
+If the specified container name already exists, `kubectl debug` will kill the
+existing container prior to starting a new one. It's legal to reuse container
+names, in which case the exited containers will continue to be reported in
+`DebugContainerStatuses`.
+
+#### Additional kubelet complexity
+
+Implementing debug requires no changes to the Container Runtime Interface as
+it's the same operation as creating a regular container. The majority of the
+complexity for this feature falls to the kubelet, but it's simplified by the
+lack of configuration and restart policy. The kubelet's additional
+responsibilities include:
+
+1.  Start or restart a debug container in a pod when requested.
+1.  Don't kill the debug container while the pod is alive.
+1.  Report status on the debug container.
+1.  Stop and delete the container when the pod is deleted.
+1.  Preserve and report state of exitted debug containers
+
+Of these, preventing `SyncPod()` from killing the container is the most complex,
+but most of this complexity has already been implemented by *init containers*.
+For *debug containers* we need only amend `computePodContainerChanges()` to
+ignore containers labeled as debug.
+
+The simplest way to preserve state of debug containers is to exempt them from
+garbage collection while the pod exists. Number of debug containers are limited
+because they are started manually and are never restarted.
```

Comment:
It should be noted that this may cause resource problems on the node, causing pods to be evicted. I think it's a reasonable tradeoff to delete debug container instances when the node is under resource pressure.